### PR TITLE
Convert only to float dtype at `dict_to_structured_array`

### PR DESCRIPTION
**Pull request by dachengx**
Otherwise, the structured array can not be easily concatenated, as in https://numpy.org/doc/stable/reference/generated/numpy.concatenate.html.

### CODE REVIEW

File: inference_interface.py [hunk 1]
```diff
@@ -243,7 +243,7 @@ def dict_to_structured_array(d):
     Return structured array with names equal to sorted(keys of d),
```
Comment (kdund) [inline inference_interface.py:245]: perhaps also just add in the docstring that we enforce float?

Comment (dachengx) [reply inference_interface.py:245]: Yes.

**Comment by kdund** [comment]
Hm, I think for me the structured arrays concatenate OK without? 
Or I'm not understanding when it fails: 

```
import numpy as np
import inference_interface as ii
dtypes = [("x",float),("y",int)]
a = np.zeros(1,dtype = dtypes)
b = np.zeros(1,dtype = dtypes)
a["x"] = 1.
a["y"] =2
b["x"] = np.pi
b["y"] = 10
zz = [ii.structured_array_to_dict(z) for z in [a,b]]
zzz = [ii.dict_to_structured_array(z) for z in zz]
zzzz = np.concatenate(zzz)
zzzz["x"], zzzz["y"]
```

**Comment by dachengx** [comment]
It is like this:
```
x = np.ones(1, dtype=[('a', int), ('b', float)])
y = np.zeros(1, dtype=[('a', float), ('b', float)])

np.concatenate([x, y])
```

when using the `null` hypothesis, the rate multiplier will be int, but in the `free` hypothesis, it is float.

**Comment by kdund** [comment]
I feel like also for the null hypothesis we should set it to 0.0? 
How does the code do it now?

**Comment by dachengx** [comment]
> I feel like also for the null hypothesis we should set it to 0.0? How does the code do it now?

Ah, yes. You are correct.

**Comment by kdund** [comment]
we can keep it as an optional thing, I guess? enforce_float_value defaulting to False?

**Comment by dachengx** [comment]
> I feel like also for the null hypothesis we should set it to 0.0? How does the code do it now?

But sometimes if we specify some `extra_args` as int and float. Like
```
extra_args: [{'a': [0.0, 1, 2]}]
```

then the error will happen.

**Comment by kdund** [comment]
Each column should be consistent I think? They're generated from the same code after all. 

So e.g. we specify WIMP masses as ints (I know, very silly, mea culpa) 
but we do that for _all_ toyMCs so it is fine?

**Comment by dachengx** [comment]
> Each column should be consistent I think? They're generated from the same code after all.
> 
> So e.g. we specify WIMP masses as ints (I know, very silly, mea culpa) but we do that for _all_ toyMCs so it is fine?

In this case:

```
extra_args: [{'a': [0.0, 1, 2]}]
```

```
{'a': 0.0}, {'a': 1}, {'a': 2}
```

will be stored in separate arrays. A similar thing happens to `generate_args`.

**Comment by kdund** [comment]
really? Will we store an array for a certain key? and then how do you convert that?

**Comment by kdund** [comment]
perversely, even this works for me: 
image

**Comment by dachengx** [comment]
> perversely, even this works for me: image

But it is an object array. I am preparing an example of the error phenomena.

**Comment by kdund** [comment]
thanks :)

**Comment by dachengx** [comment]
I do not know why I can not reproduce the error seen last time. :) Let's reopen this if needed in the future.

**Comment by dachengx** [comment]
OK, I can see one example:

at https://github.com/XENONnT/binference/blob/e9b20f626c4ed45b150faf27c505a013e794f3cc/binference/likelihoods/ll_nt_from_config.py#L542 we converted the ancillary measurement dictionary into numpy array.

```
dict_to_structured_array({'x_rate_multiplier': 0, 'y_rate_multiplier': 0.01})
```
gives you
```
array([(0, 0.01)],
      dtype=[('x_rate_multiplier', '<i8'), ('y_rate_multiplier', '<f8')])
```
and
```
dict_to_structured_array({'x_rate_multiplier': 0.01, 'y_rate_multiplier': 0.01})
```
gives you
```
array([(0.01, 0.01)],
      dtype=[('x_rate_multiplier', '<i8'), ('y_rate_multiplier', '<f8')])
```

But when we want to concatenate multiple ancillary measurements from each toy, we can not concatenate these two arrays.

**Comment by dachengx** [comment]
So the issue is originally from https://github.com/XENONnT/binference/blob/e9b20f626c4ed45b150faf27c505a013e794f3cc/binference/likelihoods/ll_nt_from_config.py#L404.

**Comment by dachengx** [comment]
@kdund how about this PR?

**Comment by kdund** [comment]
It is a hacky solution... feels like a numpy deficiency rather than an error here fundamentally, right? 

I am fine with going with this provisionally, but I do not think it is very pretty :)

**Comment by kdund** [comment]
@dachengx

**Comment by dachengx** [comment]
I think numpy indeed is the origin. But i guess this hacky thing will not lead to any potential error.

**Comment by kdund** [comment]
lets see :)

**Review by kdund — APPROVED** [review]
lets go for all-floats for now

**Pull request merged**